FDK17プロジェクト/コード/03.サウンド/CSound.cs

```diff
@@ -574,27 +579,103 @@ public double db再生速度
+	    ///
+	    /// It will in the near future also be used for adjustments based on
+	    /// loudness metadata, either when SONGVOL is unavailable or when loudness
+	    /// metadata is prioritized over it.
+	    /// </summary>
+	    public int Gain
+	    {
+	        private get => _gain;
+	        set
+	        {
+	            if (_gain == value)
+	            {
+	                return;
+	            }
+
+	            _gain = value;
+	            SetVolume();
+	        }
+	    }
+
+        /// <summary>
+        /// AutomationLevel is applied "second" to the audio data, much as in a
+        /// physical or sofware mixer and its channel level. Before this Gain is
+        /// applied, and after this the mixing group level is applied.
+        ///
+        /// This is currently used only for automated fade in and out as is the
+        /// case right now for the song selection screen background music fade
+        /// in and fade out.
+        /// </summary>
+	    public int AutomationLevel
+	    {
+	        get => _automationLevel;
+	        set
+	        {
+	            if (_automationLevel == value)
+	            {
+	                return;
+	            }
+
+	            _automationLevel = value;
+	            SetVolume();
+	        }
+	    }
+
+        /// <summary>
+        /// GroupLevel is applied "third" to the audio data, much as in the sub
+        /// mixer groups of a physical or software mixer. Before this both the
+        /// Gain and AutomationLevel are applied, and after this the audio
+        /// flows into the audio subsystem for mixing and output based on the
+        /// master volume.
+        ///
+        /// This is currently automatically managed for each sound based on the
+        /// configured and dynamically adjustable sound group levels for each of
+        /// sound effects, voice, song preview, and song playback.
+        ///
+        /// See the SoundGroupLevelController and related classes for more.
+        /// </summary>
+	    public int GroupLevel
+	    {
+	        private get => _groupLevel;
+	        set
+	        {
+	            if (_groupLevel == value)
+	            {
+	                return;
+	            }
+
+	            _groupLevel = value;
+	            SetVolume();
+	        }
+	    }
+
+	    private void SetVolume()
+	    {
+	        // 2018-08-18 twopointzero: For this version we will continue to mix integers via multiplication and use linear integer volume.
+	        // An upcoming revision will move as many values as possible to floating point logarithmic levels and use addition/subtraction.
+	        n音量 = (int) Math.Round(MixLinearLevelsLinearly(MixLinearLevelsLinearly(Gain, AutomationLevel), GroupLevel));
+	    }
+
+        private static double MixLinearLevelsLinearly(double left, double right)
+        {
+            return left * right * 0.01;
+        }
+
 		/// <summary>
 		/// <para>0:最小～100:原音</para>
 		/// </summary>
-		public int n音量
+		private int n音量
 		{
-			get
```

Comment:
With the improved sound level management in place, especially after separating out the AutomationLevel, no code needs to read back the volume so the getter has been removed.